sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -161,11 +169,17 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]]
   }
 
   def getTagValue[T](tag: TreeNodeTag[T]): Option[T] = {
-    tags.get(tag).map(_.asInstanceOf[T])
+    if (_tags eq null) {
```

Comment:
I think it makes sense to add a comment clarifying the intent of using both `_tags` and `tags` here -- "To avoid initializing the `tags` when `getTagValue` is called on a TreeNode without any tags set"